Adding Git log chapter (#2)
woojiahao/git-guide: nn-git-log/README.md

# Git Log

> Now that you know how to do the basics in git and started working on the repo, a few commits in and you might ask yourslef, how do we visualize these commits and changes we have made in the repository?  

Reviewer: - **Now that you know the basics of Git and have started working on the repo, you might ask...**
- yourslef - yourself

> Now that you know how to do the basics in git and started working on the repo, a few commits in and you might ask yourslef, how do we visualize these commits and changes we have made in the repository?  

Reviewer: ```suggestion
Now that you know the basics of Git and have started working on the repository with a few commits, you might ask yourself, "How do we visualize these commits and changes we have made in the repository?"  
```

Author: changed in commits

> That is where this chapter, and the subsequent, `Git Diffs` come in. These commands give you a way to visualize the commits and changes made in your git respository.  

Reviewer: I believe that we don't have to create a new root chapter. We could probably move them into `07-advanced-git/` and maybe have them as separate folders instead? So something like `07-01-git-log/` and house all of the resources within that folder. I can edit the home `README` to include the subsections instead and have it link to each `README` separately.

Alternatively, we could just house everything under the same document and have one giant `README` and instead store the resources in separate sub-folders under `res/` like `res/git-log/`. 

Would like your thoughts on this @chuanhao01

Author: > I believe that we don't have to create a new root chapter. We could probably move them into `07-advanced-git/` and maybe have them as separate folders instead? So something like `07-01-git-log/` and house all of the resources within that folder. I can edit the home `README` to include the subsections instead and have it link to each `README` separately.

I think this makes more sense in grouping these chapters together since its more on advanced git commands. Although I think they should still have their own `res` folder, this way we don't end up with all the images together in one place.  

> Alternatively, we could just house everything under the same document and have one giant `README` and instead store the resources in separate sub-folders under `res/` like `res/git-log/`.

In line with the reason above, it might also make development on additional chapters harder, since everything is in one big `README.md`.  

We could add a summary `README.md`  to give new readers a sense of what advanced chapters might be better for them. So, in the end, the file structure would look something like this.  
```
.
│
├── ...
├── 07-advanced-git
│   ├── 07-01-git-log
│   │   ├── res
│   │   │   └── ...
│   │   └── README.md
│   ├── 07-02-...
│   ├── 07-03-...
│   └── README.md
└── ...
```

Edit:
Looks like there is a lot more to read, will read through at night when I have more time.
Will update this edit when done.

Reviewer: Yeap, I was referring to exactly that in the simple diagram you included :+1:

Reviewer: Resolving this as the changes have been reflected :+1:

> To explain simply, the command `git log` and its corresponding optional flags lets you look at history of commits and how the link up in your respository. At the end of the chapter there will be links to additional resources, as well as more visual GUI tools you can use to look at the log of the respository, but I still suggest reading and learning this chapter.  

Reviewer: Could probably trim the word content by merge line 7 and line 9 together. For e.g.

> ... At the ... but I still suggest reading and understanding how `git log` works as these tools still rely on the basic `git log` to generate their graphs. Having an understanding of this command gives you an intuition for how these tools work in general. Additionally, as the command comes with Git, there is no additional setup required to start using `git log`.

Author: ```suggestion
**TL|DR:** The command `git log` and its corresponding optional flags lets you look at history of commits and how they link up in your respository.  

At the end of the chapter there will be links to additional resources, as well as more visual GUI tools you can use to look at the log of the respository, but I still suggest reading and understanding how `git log` works as these tools still rely on the basic `git log` to generate their graphs. Having an understanding of this command gives you an intuition for how these tools work in general. Additionally, as the command comes with Git, there is no additional setup required to start using `git log`. 
```

I think I want to keep l part of line 7 as the first part was a TLDR.
Will combine the rest of line 7 and 9 into 1 line to be more concise
Done in the new commits

Reviewer: Will make the comment on the latest changes. Resolving this for now :+1:

This is because these tools still do rely on the git commands to generate their graphs, and understanding this command kind of gives you the 'feel' for it in general. Lastly, as this command comes with git, using the command does not require any additional setup other than having git.

Note:
This chapter serves to show you and explain the common use cases for the command. For a more detailed explaination, look at the addtional resources or the documentation of the command.

## Project set up so-far

Let's say the project you are working on is a full-stack web application, with a frontend and backend.  
And you have decided to split the work to frontend and backend work, with respective features branches being worked on.

It may look something like this:  
```

> Github

Reviewer: Are these the branches that each teammate has? Would it be better for them to indicate explicitly that that is what the code block is for, so they don't get confused?

Author: ```
Github branches:
master, start-server (This will be remote branches for you)

Teammate A's branches (On your local machine):
master, origin/master, backend/start-server, origin/backend/start-server, frontend/landing

Teammate B's branches:
backend/start-server (His local branch connected to github)
```

Tried to be more explicit, but I'm not sure if this helps
pushed in new commits

Reviewer: Yeap, these changes work :+1: Much clearer as to what we are seeing.

Minor changes to be added but will add them in the latest comment for review

origin/master, master

Teammate A
frontend/landing

Teammate B
backend/start-server
```

## Looking at the small picture

**Note: Assume you are teammate A**

Let's say at this point you have added some code and commits to the project and want to see what commits you had done.  

> This could be because, you wanted to see where you left off yesterday or how features have or have not been done. (The later reason is when you are working on a larger feature)

Reviewer: ```suggestion
This could be because, you wanted to see where you left off yesterday or how many features have been completed. (The latter reason is when you are working on a larger feature)
```

Author: Oops I didn't proofread this, must have been too late haha :D.
changes in commits

Reviewer: :joy: No worries, resolving this :+1:

### Basic git log

Well to look at all the commits made on your current branch you can do:
```bash
git log
```
![Basic git log of local branch](res/small_picture_1.png)

From this screenshot you can see that although we can see all the commits linked to our current branch, there is a bit too much information at once.  
For example information like when the commit was made or who made it is not really important when trying to look at the bigger picture.  
This is also not practical when looking at projects with 10s or 100s of commits, as we would be lost in the information

> ### Flag oneline

Reviewer: We could branch this section into 
```
### Flags
#### `--online`
```

Author: changed the branches in commits

Reviewer: Resolving this as changes are in order :+1:

Luckily there are special flags we can use to format the output.  
If we add the flag, `--oneline`, git will shorten each commit to oneline, only preserving the important information.  
Looking at the man page for this command:  
```bash
Documentation:
--oneline
This is a shorthand for "--pretty=oneline --abbrev-commit" used together.
--abbrev-commit
Instead of showing the full 40-byte hexadecimal commit object name, show only a partial prefix. Non default number of digits can be specified with "--abbrev=<n>" (which also modifies diff output, if it is displayed).
```
Thus when running the command.
```bash
git log --oneline
```
![Git log with only oneline](res/small_picture_2.png)

And boom, this was what we wanted at the start. Now we can see that we have added the landing page and also changed the extension of the landing page.  

> But there is something weird here, we can also see that the `origin/master` commit together with the commits of our branch. This is because as we know, branches have to start off(be based off) somewhere. As such, we can make a pretty good guess that this `frontend/landing` was made off the master branch.  

Reviewer: I think this may be slightly confusing for some. Would it be better if we rephrased it as:

> But there is something weird here... We can see that there is a commit tagged with `origin/master`, along with the commits of our current branch, `frontend/landing`. As we have established previously in chapter 6, all branches start off somewhere. In this example, the branch `frontend/landing` branches out from the `origin/master` branch. As such, `git log` makes a reference to this.

Author: made the change in the new commits

Reviewer: Resolving this :+1:

Luckily, we don't have to guess as there is a flag which will show us the relationship between these commits.  

### Flag graph

> To make git log show the relationships between commits, we just have to add the `--graph` flag. This flag shows a 'graph' of the commits, which for us means the relationship the commits have.  

Reviewer: Would be better to elaborate on what this `relationship` means. Is it the relationship between branches? Or something else?

Author: Relationships between commits
I thought it should be relationships between commits and the branches implicitly?
Cause from my understanding, it was that if u track the branches commits, you can also 'see' how the branches link up/

Reviewer: Okay, reading over it again, it makes sense. Would have to improve the phrasing of the conclusion from the previous section then :ok_hand: (Opening new review on latest commits)

Combining this with `--oneline` flag we showed above, makes for a very clean and concise git log.  

> (I will leave it to the reader to try out what `--graph` flag does on its own)

Reviewer: ```suggestion
Try this out yourself and see what happens!
```

Author: changed in commits

```bash
git log --oneline --graph
```
![Git log with oneline and graph](res/small_picture_3.png)

## Looking at the big picture
Now that you know how to look at commits tagged to your branch, you might also want to see how the entire repository might look like.  
Luckily we have a flag to look at the log of the entire repository.  

### Flag all

> Using the `--all` flag, this will show the git log, including all the commits from every branch you might have in the respository.  

Reviewer: ```suggestion
The `--all` flag shows the `git log`; including all the commits from every branch you have in the repository.  
```

Author: changed in commits

> We will also be combining this flag with all previous flags to get the best overview of a project. (In my opnion)

Reviewer: ```suggestion
We will also be combining this flag with all previous flags to get the best overview of a project. (In our opinion)
```

Author: change in commits

```bash
git log --oneline --graph --all
```
![Git log with oneline graph and all](res/big_picture_1.png)

> Now we can see that our frontend landing page feature has already been merged down to master. We can also see that after git fetch(To fetch any changes from github), our teammate B had already made more commits on his feature and our local `backend/start-server` branch is behind.  

Reviewer: We could use a short chapter on `git fetch`. The current method introduced is just `git pull` since I didn't want to confuse them in the core chapters but `git fetch upstream master && git merge upstream/master` is definitely still better

Author: Ah didn't realise this, I thought since u did pull, the concept of remote branches was introduced.
Should include in the project timeline.

Reviewer: Yeap, adding this to the Projects board as a To Do task :+1:

> Now we can see that our frontend landing page feature has already been merged down to master. We can also see that after git fetch(To fetch any changes from github), our teammate B had already made more commits on his feature and our local `backend/start-server` branch is behind.  

Reviewer: ```suggestion
Now we can see that our frontend landing page feature has already been merged down to master. We can also see that after git fetch (To fetch any changes from GitHub), teammate B had already made more commits on his feature branch and the local `backend/start-server` branch is behind.  
```

Author: Man, I must have been very tired.
Will read and look through before asking for a review next time.
change in commits

Reviewer: No worries at all :joy: :+1: That's what the PR system is for innit :stuck_out_tongue:

> Now we can see that our frontend landing page feature has already been merged down to master. We can also see that after git fetch(To fetch any changes from github), our teammate B had already made more commits on his feature and our local `backend/start-server` branch is behind.  

Reviewer: Perhaps discuss how the lines on the left hand side of `git log` show the merge history as well? So that this doesn't suddenly introduce a new concept of having the frontend feature branch be merged with the `master` branch. Also, it might be good to specify that the merge has occurred in the local `master` branch and not the remote. I will be going into the naming convention of branching too so that this isn't a new bombshell of knowledge on them

Author: > Perhaps discuss how the lines on the left hand side of `git log` show the merge history as well? So that this doesn't suddenly introduce a new concept of having the frontend feature branch be merged with the `master` branch.

Agreed.
Initially, I was thinking of adding a section on how to read and relate the left-hand side to the conceptual diagrams of git branching but thought that was too much. I'm still on the french about this.

> Also, it might be good to specify that the merge has occurred in the local `master` branch and not the remote.

Will add more explanations to make it more explicit.

> I will be going into the naming convention of branching too so that this isn't a new bombshell of knowledge on them

Not sure if I followed the naming convention on this haha XD.

Reviewer: > Initially, I was thinking of adding a section on how to read and relate the left-hand side to the conceptual diagrams of git branching but thought that was too much. I'm still on the french about this.

I think you can connect this portion with the image of the GitHub visualisation tool. GitHub's tool is pretty clear to understand.

> Not sure if I followed the naming convention on this haha XD.

I believe the convention on GitHub is just `<username>:<branch>`

Reviewer: Resolving this for now but will bring this up again in the new review :+1:

## Wrapping up

> To end off, the `git log` command actually has a lot more features and flags which are a bit more nuanced and situationally. What was shown here are the most common and applicable features and flags which I hope will be usefully to you.

Reviewer: ```suggestion
To end off, the `git log` command actually has a lot more features and flags which are a bit more nuanced and situational. This chapter demonstrates the most commonly used features and flags which we hope will be usefully to you.
```

Author: change in commits

As mentioned above, below are additional resources:  

> [More features of git log](https://www.atlassian.com/git/tutorials/git-log)

Reviewer: Set this to be a list instead as now the text are all on the same line

```md
[]()

[]()
```

If you want the content to be on the new lines

Author: Not sure what change you want here, but yea I saw the pdf output being weird.
Definitely want it to be on a new line.

Reviewer: Could make this a list so like 
```md
- []()
- []()
```

The other stuff I sent was just formatting. Will open another review for this :+1:

[git log documentation](https://git-scm.com/docs/git-log)

Additional GUI applications that can help you visualise the git log:  
[VSCode extension (Git Graph)](https://marketplace.visualstudio.com/items?itemName=mhutchie.git-graph)
![Git graph looks like](res/git_graph_vscode.png)

> On Github:

Reviewer: Could be set to a subheading instead so it shows up when `pandoc` renders the PDF document for ease of reference

Author: Also not sure here, new to pandoc so I'm not so sure of how to set up so the conversion goes smoothly.
Tried referencing your previous readme although I'm not sure what you did.

Reviewer: Just a regular subheading in markdown will do :+1: Will open another issue

```bash
git log
```
![Basic git log of local branch](res/small_picture_1.png)

From this screenshot you can see that although we can see all the commits linked to our current branch, there is a bit too much information at once.  
For example information like when the commit was made or who made it is not really important when trying to look at the bigger picture.  
This is also not practical when looking at projects with 10s or 100s of commits, as we would be lost in the information

### Flag oneline

Luckily there are special flags we can use to format the output.  
If we add the flag, `--oneline`, git will shorten each commit to oneline, only preserving the important information.  
Looking at the man page for this command:  
```bash
Documentation:
--oneline
This is a shorthand for "--pretty=oneline --abbrev-commit" used together.
--abbrev-commit
Instead of showing the full 40-byte hexadecimal commit object name, show only a partial prefix. Non default number of digits can be specified with "--abbrev=<n>" (which also modifies diff output, if it is displayed).
```
Thus when running the command.
```bash
git log --oneline
```

> ![Git log with only oneline](res/small_picture_2.png)

Author: Was wondering if adding the documentation here is good.
Documentation is taken from the man page.
[Man Page](https://git-scm.com/docs/git-log)

Honestly, I think it should not be included, but wanted to get your opinion on it @woojiahao

Reviewer: I think that we can include the documentation for flags that are important to the discussion. For instance, in the commit above about `git log` have a simple description of `--oneline` does provide some help into showing ways to improve the use of the command. I don't think we need to include all of the possible flags. A link to the main page is sufficient for them to read more :thinking:

Reviewer: Resolving this
